### ISSUE TITLE
recommended way to extend MsfClient?

### ISSUE THREAD

**Issue by arcanor** [alvyxaz/barebones-masterserver#91]
I'm working on implementing my own new custom module. I think I've got the server side done, but obviously I can't test it until I have a client to work with the new module. But I'm not sure how I should be extending my client in accordance with MSF. When I look at MsfClient.cs, I see all the various client classes of MSF in there. But there's no hook for me to add more. Maybe we need something like this:
`public MsfCustomClient CustomClient { get; private set; }`
... along with the accompanying instantiation and connection setting below.

Or did you have a different way in mind for adding custom client classes to MSF?

**Comment by arcanor** [comment]
For example, here's how I've done things for now (but it's a hack, since I'm directly editing MsfClient.cs)...
`	using Barebones.Networking;

	namespace Barebones.MasterServer
	{
		public class MsfClient : MsfBaseClient
		{
			public MsfRoomsClient Rooms { get; private set; }

			public MsfSpawnersClient Spawners { get; private set; }

			public MsfMatchmakerClient Matchmaker { get; private set; }

			public MsfAuthClient Auth { get; private set; }

			public MsfChatClient Chat { get; private set; }

			public MsfLobbiesClient Lobbies { get; private set; }

			public MsfProfilesClient Profiles { get; private set; }

			// custom client class declarations:
			public Arcanoria.Revelations.MsfToonSelectionClient ToonSelection { get; private set; }

			public MsfClient(IClientSocket connection) : base(connection)
			{
				Rooms = new MsfRoomsClient(connection);
				Spawners = new MsfSpawnersClient(connection);
				Matchmaker = new MsfMatchmakerClient(connection);
				Auth = new MsfAuthClient(connection);
				Chat = new MsfChatClient(connection);
				Lobbies = new MsfLobbiesClient(connection);
				Profiles = new MsfProfilesClient(connection);

				// custom client class instantiations:
				ToonSelection = new Arcanoria.Revelations.MsfToonSelectionClient(connection);
			}
		}
	}
`

I feel there should be a better way than this. Recommendations?

**Comment by alvyxaz** [comment]
A very good question!

At the moment, everything that goes under the `Msf.*` is part of the **Msf API**, so you should add things to `Msf.*` only if you want to extend the API and make the changes part of the official release.

For your custom stuff, you don't need to use or change the `Msf` object. You could make one of your own, like: 

``` C#
/// <summary>
    /// Custom static class for "quick access", similar to `Msf.*`
    /// </summary>
    public static class MyCustom
    {
        public Arcanoria.Revelations.MsfToonSelectionClient ToonSelection { get; private set; }

        /// <summary>
        /// Initialize "client modules" as they are initialized in 'Msf' object
        /// </summary>
        static MyCustom()
        {
            // Use the `Msf.Connection` to pass it to your custom module
            ToonSelection = new Arcanoria.Revelations.MsfToonSelectionClient(Msf.Connection);
        }

        /// <summary>
        /// Or you could just make the necessary calls directly
        /// </summary>
        public static SelectToon()
        {
            Msf.Connection.SendMessage(5, (status, response) =>
            {
                
            });
        }
    }
```

Another solution could be to add a few generic methods to API, like:

``` C#
Msf.CustomModules.Add(new CustomModule());
Msf.CustomModules.Get<CustomModule>();
```

But it would introduce a new problem - we would always need to make sure that "module clients" are added before trying to get them. This wouldn't normally be a problem, but since many people do stuff in unity's **Awake/Start** methods - you can't guarantee the order of execution, unless you do additional "tying things together", which is too much of a bother for something that's only a minor syntactic sugar.

**Comment by arcanor** [comment]
Thanks for your quick response. :)

I had considered creating my own static singleton class (similar to what you suggested above), but I was concerned that I would lose the advantages of the already centrally-managed MsfClient class. Specifically, I hoped to ensure that my custom "module client code" class is instantiated alongside all the other MsfClient member classes, with a common place for storing all client class references. I guess it's not that important though, so I should probably create my own singleton and avoid the hack to allow easier merging in the future.

Maybe I'm being obtuse, or maybe it's a matter of dedicating more time to studying the intricacies of what's already inside MSF. But I'm finding that it's not always obvious where to draw the line between the framework and my own custom extensions. MSF does so many things, but it doesn't always provide "hooks" to extend its basic demonstration functionality natively. Instead, I need to change my thinking and develop my own groups/clusters of classes to re-implement MSF patterns in parallel. I am finding that this requires a deep understanding of how MSF works.

I guess what I'm saying is this: Everything provided in the MSF demos is built from "inside" the framework, i.e. embedded in Msf.* classes, using internal instantiations, hooked into the demo UIs, etc. What I'm finding is lacking are concrete examples of how to extend the framework from "outside", which I feel is kinda "the hard way" ;). I'd like to see examples of how to customize in various ways, without them being tied directly into Msf.* classes for initialization, etc. Template classes that can be copied, renamed, and directly modified for use in a specific custom project.

I think the demos already do quite a lot, and I thank you for providing those. I'm just finding it difficult sometimes to extrapolate those lessons learned into my own working custom extensions.

**Comment by arcanor** [comment]
I should also add that I've already mostly got my custom MSF module working in my own project, so I expect this is something I can help to document, along with the rest of the community, in the coming weeks/months. However, as you know, it often takes even more time to document and support something, than to produce a working version in the first place. :)

I believe that between me, you, and the rest of the community, we will find a way to create and share some good **Reusable Template Classes for extending MSF**, along with documentation for how they work. It's a big task, but we're a growing community with many members using MSF in many different ways, so I am hopeful.

MSF is already fantastic. This will make it even better, in my opinion. :)

**Comment by alvyxaz** [comment]
@arcanor Thanks a lot for being so supporting!

And yeah, I can understand that it feels very daunting when you want to do something **properly**, and there's a bunch of different ways to do something, so you can't be sure if you're using the right one.

I agree, that the process of working with MSF should be more streamlined. I think it would be a good idea to create an issue/discussion for this, where we could "design" the workflow. 

For example, we could have a list, like

Workflow:
* Get master server working
* Add/Use some of the modules
* Extend existing module
* Write a custom module from scratch

Then, users could request what they're missing, for example, "how to change the state of the peer", and we could add it to the "workflow". When a list like this is complete, it would be quite easy to fill in the gaps in the documentation :)

Let me know what you think about this. If you think this could work, I'll create an issue/discussion :)

**Comment by arcanor** [comment]
Yes, I think the documentation would benefit from and explanation of the concept of MSF Workflow as you outlined above. And those workflow bullets could hyperlink to more specific documents for each step.
